applications/crossbar/src/modules/cb_allotments.erl

```diff
@@ -0,0 +1,326 @@
+%% @end
+%%--------------------------------------------------------------------
+-spec is_allowed(cb_context:context()) -> boolean().
+is_allowed(Context) ->
+    AccountId = cb_context:account_id(Context),
+    AuthAccountId = cb_context:auth_account_id(Context),
+    IsSystemAdmin = wh_util:is_system_admin(AuthAccountId),
+    {'ok', MasterAccount} = whapps_util:get_master_account_id(),
+    case wh_services:find_reseller_id(AccountId) of
+        AuthAccountId ->
+            lager:debug("allowing reseller to update allotments"),
+            'true';
+        MasterAccount ->
+            lager:debug("allowing direct account to update allotments"),
+            'true';
+        _Else when IsSystemAdmin ->
+            lager:debug("allowing system admin to update allotments"),
+            'true';
+        _Else ->
+            lager:debug("sub-accounts of non-master resellers must contact the reseller to change their allotments"),
+            'false'
+    end.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec maybe_handle_load_failure(cb_context:context()) ->
+                                       cb_context:context().
+-spec maybe_handle_load_failure(cb_context:context(), pos_integer()) ->
+                                       cb_context:context().
+maybe_handle_load_failure(Context) ->
+    maybe_handle_load_failure(Context, cb_context:resp_error_code(Context)).
+
+maybe_handle_load_failure(Context, 404) ->
+    Data = cb_context:req_data(Context),
+    NewLimits = wh_json:from_list([{<<"pvt_type">>, ?PVT_TYPE}
+                                   ,{<<"_id">>, ?PVT_TYPE}
+                                  ]),
+    JObj = wh_json_schema:add_defaults(wh_json:merge_jobjs(NewLimits, wh_json:public_fields(Data))
+                                       ,<<"limits">>
+                                      ),
+
+    cb_context:setters(Context
+                       ,[{fun cb_context:set_resp_status/2, 'success'}
+                         ,{fun cb_context:set_resp_data/2, wh_json:public_fields(JObj)}
+                         ,{fun cb_context:set_doc/2, crossbar_doc:update_pvt_parameters(JObj, Context)}
+                        ]);
+maybe_handle_load_failure(Context, _RespCode) -> Context.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec update_allotments(cb_context:context()) -> cb_context:context().
+update_allotments(Context) ->
+    Doc = cb_context:doc(Context),
+    Allotments = cb_context:req_data(Context),
+    NewDoc = wh_json:set_value(?PVT_ALLOTMENTS, Allotments, Doc),
+    Context1 = crossbar_doc:save(cb_context:set_doc(Context, NewDoc)),
+    cb_context:set_resp_data(Context1, Allotments).
+
+-spec cycle_start(ne_binary(), wh_datetime()) -> gregorian_seconds().
```

Comment:
This all looks similar to code in `j5_allotments`; let's break this out into its own core file `core/kazoo_documents/src/kzd_allotment.erl` The building of the allotment_consumption docs should be moved here too (but that can be filed as a ticket - not necessary for this PR).